src/javascript/campaign-url-builder/components/problematic.js

```diff
@@ -0,0 +1,162 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * This file manages "problematic" URLs. For instance, if users are
+ * creating links to Play Store URLs, they should user the Google Play
+ * Url Builder (https://goo.gl/qycyHd) instead of the GA devtools app.
+ *
+ * See b/69681865 for more details.
+ */
+
+import React from 'react';
+import {isRegExp} from 'lodash';
+
+
+/**
+ * Create a function that, given a RegExp instance, returns a function
+ * that tests a url against that regex.
+ *
+ * @param  {RegExp} regexp compiled regular expression object used in
+ * the test
+ * @return {Function} function that tests the given URL
+ * against the regex
+ */
+const matchesRegexp = regexp => url => regexp.test(url);
+
+/**
+ * Given either a RegExp or a pattern, return a function that tests if
+ * a given URL matches that pattern.
+ *
+ * @param  {string|Regexp} pattern to use in the returned testing
+ * function. If the pattern is a string, it is compiled (case-insensitive)
+ * to a Regexp
+ * @return {Function} function that tests if a URL matches
+ * the given regexp
+ */
+const matches = pattern => isRegExp(pattern) ?
+  matchesRegexp(pattern) :
+  matchesRegexp(new RegExp(pattern, 'i'));
+
+
+/**
+ * Given a domain, create a function that returns true if the domain
+ * matches a given URL. The domain is matched exactly, and can optionally
+ * be prepended with http:// or https://.
+ *
+ * @param {string} domain the domain to match. Can be a regex pattern, but all
+ * . characters are unconditionally escaped
+ *
+ * @return {Function} a function that returns true if the
+ * domain matches the given url exactly
+ */
+const domainMatches = domain =>
+  matches('^(?:https?://)?(:?' + domain.replace('.', '\\.') + ')(?:$|[/?#])');
+
+/**
+ * Given a domain suffix, create a function that returns true if the
+ * domain matches a given URL. The domain is matched as a suffix; that
+ * is, the domain "google.com" will match "play.google.com",
+ * "mail.google.com", etc, as well as just "google.com".
+ *
+ * @param {string} suffix Suffix of the domain to match. Can be a pattern,
+ * but all . characters are unconditionally escaped
+ *
+ * @return {Function} a function that returns true if the
+ * domain matches the given url as a suffix
+ */
+const domainSuffix = suffix =>
+  domainMatches(`(?:[a-z0-9-]+.)*(?:${suffix})`);
+
+const googlePlayBuilderUrl = 'https://developers.google.com/analytics/devguides/collection/android/v4/campaigns#google-play-url-builder';
+const itunesStoreBuilderUrl = 'https://developers.google.com/analytics/devguides/collection/ios/v3/campaigns#url-builder';
+
+// Many fields in the badlist require a function with the signature
+// "url => thing". Use literal to make a function that just returns a
+// literal value
+const literal = thing => url => thing;
```

Comment:
This seems unnecessarily complex to me. Why not just use a `switch` statement?

While a `switch` statement would definitely be less clever, I think it would be clearer and not depend on a closure to do a pretty simple thing.

Reply:
Yeah, I went back and forth on this. I liked how this made the renderProblematic function more simple, and the BADLIST more explicit, but I can revert it

Reply:
Oh, I also liked how it removed a conditional in `renderProblematic`; as a rule I like to avoid conditionals, if possible. But still, fixed.